tests/common/gcp_type/firewall_rule_test.py

```diff
@@ -25,6 +25,221 @@
+      rule = firewall_rule.FirewallRule.from_json(json_string)
+      self.assertTrue(rule.validate())
+
+    @parameterized.parameterized.expand([
+        (
+            {
+                'kind': 'compute#firewall',
+                'id': '8',
+                'creationTimestamp': '2017-05-01T22:08:53.399-07:00',
+                'name': 'default',
+                'description': '',
+                'network': 'network name',
+                'priority': 1000,
+                'sourceRanges': ['0.0.0.0/0'],
+                'allowed': [
+                    {
+                        'IPProtocol': 'tcp',
+                        'ports': ['22']
+                    }
+                ],
+                'direction': 'EGRESS',
+                'selfLink': 'https:// insert link here'
+            },
+            firewall_rule.InvalidFirewallRuleError,
+        ),
+        (
+            {
+                'kind': 'compute#firewall',
+                'id': '8',
+                'creationTimestamp': '2017-05-01T22:08:53.399-07:00',
+                'description': '',
+                'network': 'network name',
+                'priority': 1000,
+                'sourceRanges': ['0.0.0.0/0'],
+                'allowed': [
+                    {
+                        'IPProtocol': 'tcp',
+                        'ports': ['22']
+                    }
+                ],
+                'direction': 'INGRESS',
+                'selfLink': 'https:// insert link here'
+            },
+            firewall_rule.InvalidFirewallRuleError,
+        ),
+        (
+            {
+                'kind': 'compute#firewall',
+                'id': '8',
+                'creationTimestamp': '2017-05-01T22:08:53.399-07:00',
+                'description': '',
+                'network': 'network name',
+                'priority': 1000,
+                'sourceRanges': ['0.0.0.0/0'],
+                'direction': 'INGRESS',
+                'selfLink': 'https:// insert link here'
+            },
+            firewall_rule.InvalidFirewallRuleError,
+        ),
+        (
+            {
+                'kind': 'compute#firewall',
+                'id': '8',
+                'creationTimestamp': '2017-05-01T22:08:53.399-07:00',
+                'name': 'default',
+                'description': '',
+                'network': 'network name',
+                'priority': -1,
+                'sourceRanges': ['0.0.0.0/0'],
+                'allowed': [
+                    {
+                        'IPProtocol': 'tcp',
+                        'ports': ['22']
+                    }
+                ],
+                'direction': 'INGRESS',
+                'selfLink': 'https:// insert link here'
+            },
+            firewall_rule.InvalidFirewallRuleError,
+        ),
+    ])
+    def test_from_json_error(self, json_dict, expected_error):
+      json_string = json.dumps(json_dict)
+      with self.assertRaises(expected_error):
```

Comment:
I recommend switching to assertRaisesRegexp and including in your test parameters the expected error message string regexp to verify the correct reason for the exception is tested. This is true for the other places assertRaises is used as well.

Reply:
Done.